### PR TITLE
Add FileStream and int64 overloads to SendPacketsElement (take 2)

### PR DESCRIPTION
**Pull request by kurtschelfthout**
See issue #25354

This is basically a continuation of what @AlexRadch did in the closed PR #27839 . I merged from master, and addressed the review comments in that PR, in particular:

@karelz 
> The API was approved as OffsetLong

and 

@tmds 

> I think it makes sense to make SendPackets throw when used with non async FileStreams. Then we can remove the workaround in SocketAsyncEventArgs.Windows.cs and remove these special cases from the tests. It has consistent cross-platform behavior and I don't see any downsides.

### CODE REVIEW

**Comment by kurtschelfthout** [comment]
Thanks @jkotas. Done. Also neglected to remove a workaround that surfaced in the earlier review.

**Comment by kurtschelfthout** [comment]
It looks like the NETFX and UWP builds are failing because build of the tests fails - those builds don't see the updated code, e.g. the new constructors. I assume there is something special I need to do here because the API changed?

File: src/System.Net.Sockets/src/Resources/Strings.resx [hunk 1]
```diff
@@ -244,4 +244,7 @@
   <data name="InvalidOperation_MultipleContinuations" xml:space="preserve">
     <value>Another continuation was already registered.</value>
   </data>
-</root>
+  <data name="net_sockets_sendpackelement_FileStreamMustBeAsync" xml:space="preserve">
+    <value>The FileStream must be opened asynchronously.</value>
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/Resources/Strings.resx]: Nit: "The FileStream must have been opened for asynchronous reading and writing."

File: src/System.Net.Sockets/src/System/Net/Sockets/SendPacketsElement.cs [hunk 2]
```diff
@@ -39,7 +43,40 @@ public SendPacketsElement(string filepath, int offset, int count, bool endOfPack
...
+        public SendPacketsElement(FileStream fileStream, long offset, int count, bool endOfPacket)
+        {
+            // We will validate if the fileStream exists on send.
+            if (fileStream == null)
+            {
+                throw new ArgumentNullException(nameof(fileStream));
+            }
+            if (!fileStream.IsAsync)
+            {
+                throw new ArgumentException(nameof(fileStream), SR.GetResourceString("net_sockets_sendpackelement_FileStreamMustBeAsync"));
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SendPacketsElement.cs]: `SR.GetResourceString("net_sockets_sendpackelement_FileStreamMustBeAsync")` should instead be `SR.net_sockets_sendpackelement_FileStreamMustBeAsync`

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SendPacketsElement.cs]: ArgumentException's arguments should be reversed; they're unfortunately not in the order you'd expect.

**Comment by kurtschelfthout** [comment]
Thanks @stephentoub, all fixed.

**Comment by rmkerr** [comment]
@kurtschelfthout 

If you're using an API that differs on netfx, you can split out the tests into a dedicated test file that is conditionally included only when targeting netcoreapp. For example, see how we include `SocketAsyncEventArgsTest.netcoreapp.cs`:
https://github.com/dotnet/corefx/blob/0ec9c9d7caac1ac378e2135620bf2f8b80e6ed86/src/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj#L44-L46

**Comment by danmosemsft** [comment]
@stephentoub it looks like this is ready for re-review.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs [hunk 3]
```diff
@@ -648,51 +652,44 @@ internal unsafe SocketError DoOperationSendPackets(Socket socket, SafeSocketHand
             // Attempt to open the files if any were given.
             if (sendPacketsElementsFileCount > 0)
             {
-                // Create arrays for streams.
-                _sendPacketsFileStreams = new FileStream[sendPacketsElementsFileCount];
-
                 // Loop through the elements attempting to open each files and get its handle.
                 int index = 0;
-                foreach (SendPacketsElement spe in sendPacketsElementsCopy)
+                // Create arrays for streams.
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Nit: this comment isn't necessary

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs [hunk 4]
```diff
@@ -648,51 +652,44 @@ internal unsafe SocketError DoOperationSendPackets(Socket socket, SafeSocketHand
...
+                    throw;
+                }
             }
 
-            if (sendPacketsElementsFileCount + sendPacketsElementsBufferCount == 0)
+            if (sendPacketsElementsFileCount + sendPacketsElementsFileStreamCount + sendPacketsElementsBufferCount == 0)
             {
                 FinishOperationSyncSuccess(0, SocketFlags.None);
                 return SocketError.Success;
             }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs]: Should this whole if block be moved to above the previous whole if block?

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        {
+            using (var stream = new FileStream(TestFileName, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, FileOptions.Asynchronous))
+            {
+                var elements = new[]
+                {
+                    new SendPacketsElement(stream, 0, 20),
+                    new SendPacketsElement(stream, s_testFileSize - 10, 10),
+                    new SendPacketsElement(stream, 0, 10),
+                    new SendPacketsElement(stream, 10, 20),
+                    new SendPacketsElement(stream, 30, 10),
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: I might just be missing it, but we should have tests that validate a SendPacketsElement[] that contains buffers, paths, and streams, verifying they were all sent and all sent in the right order.  Do we?

Comment (kurtschelfthout) [reply src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: We don't, I agree that would be valuable. Working on it.

Comment (kurtschelfthout) [reply src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Done...also checked that the bytes that are sent are actually the same, vs just the number sent.

**Comment by kurtschelfthout** [comment]
Thanks all, I've addressed all of the re-review comments.

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs [hunk 6]
```diff
@@ -461,8 +461,17 @@ private void SendPackets(SocketImplementationType type, SendPacketsElement[] ele
...
-                        Assert.Equal(expectedResut, args.SocketError);
+                        Assert.Equal(expectedResult, args.SocketError);
                         Assert.Equal(bytesExpected, args.BytesTransferred);
+                    
+                    }
+
+                    if (contentExpected != null) {
+                        // test server should just echo back, so read length expected bytes from the stream
+                        var contentActual = new byte[bytesExpected];
+                        var bytesReceived = sock.Receive(contentActual);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs]: Nit: var => int

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 7]
```diff
@@ -214,5 +181,65 @@ public void SendPacketsElement_FileStreamLargeCount_Throws(SocketImplementationT
...
+                if (element.FilePath != null) {
+                    return (int) new FileInfo(element.FilePath).Length;
+                }
+                else if (element.FileStream != null) {
+                    return (int) element.FileStream.Length;
+                }
+                throw new ArgumentException("Expected SendPacketsElement with FilePath or FileStream set.", nameof(element));
+            }
+
+            var totalCount = 0;
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Nit: var => int

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 8]
```diff
@@ -214,5 +181,65 @@ public void SendPacketsElement_FileStreamLargeCount_Throws(SocketImplementationT
...
+            foreach (var element in elements) {
+                totalCount += element.Buffer != null ? element.Count : FileCount(element);
+            }
+            var result = new byte[totalCount];
+            var resultOffset = 0L;
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Nit: var => long

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 9]
```diff
@@ -180,8 +180,61 @@ public void SendPacketsElement_FileStreamLargeCount_Throws(SocketImplementationT
...
+                var elements = new[]
+                {
+                    new SendPacketsElement(new byte[] { 5, 6, 7 }, 0, 3),
+                    new SendPacketsElement(stream, s_testFileSize - 10, 10),
+                    new SendPacketsElement(TestFileName, 0L, 10),
+                    new SendPacketsElement(stream, 10L, 20),
+                    new SendPacketsElement(TestFileName, 30, 10),
+                    new SendPacketsElement(new byte[] { 8, 9, 10 }, 0, 3),
+                };
+                var expected = GetExpectedContent(elements);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Nit: var => ?

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 10]
```diff
@@ -180,8 +180,61 @@ public void SendPacketsElement_FileStreamLargeCount_Throws(SocketImplementationT
...
+                    var elements = new[]
+                {
+                    new SendPacketsElement(new byte[] { 5, 6, 7 }, 0, 0),
+                    new SendPacketsElement(stream, s_testFileSize - 10, 10),
+                    new SendPacketsElement(stream2, s_testFileSize - 100, 10),
+                    new SendPacketsElement(TestFileName, 0L, 10),
+                    new SendPacketsElement(new byte[] { 8, 9, 10 }, 0, 1),
+                    new SendPacketsElement(TestFileName, 30, 10),
+                };
+                var expected = GetExpectedContent(elements);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Nit: var => ?

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs [hunk 11]
```diff
@@ -469,7 +469,14 @@ private void SendPackets(SocketImplementationType type, SendPacketsElement[] ele
                     if (contentExpected != null) {
                         // test server should just echo back, so read length expected bytes from the stream
                         var contentActual = new byte[bytesExpected];
-                        var bytesReceived = sock.Receive(contentActual);
+                        var bytesReceived = 0;
+                        var tries = 0;
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs]: Nit: vars => ints

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs [hunk 12]
```diff
@@ -469,7 +469,14 @@ private void SendPackets(SocketImplementationType type, SendPacketsElement[] ele
...
+                        const int MaxTries = 15;
+                        while (bytesReceived < bytesExpected && tries < NaxTries) {
+                            bytesReceived += sock.Receive(contentActual, bytesReceived, bytesExpected, SocketFlags.None);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs]: There's no guarantee that Receive returns anything larger than 1.  Why the 15 limit?

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs [hunk 13]
```diff
@@ -233,6 +233,23 @@ public void SendPacketsElement_FileStreamLargeCount_Throws(SocketImplementationT
...
+        public void SendPacketsElement_FileStreamMultiPartMixed_MultipleWholeFiles_Success(SocketImplementationType type) {
+            using (var stream = new FileStream(TestFileName, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, FileOptions.Asynchronous)) {
+                var elements = new[]
+                {
+                    new SendPacketsElement(stream, 0L, 0),
+                    new SendPacketsElement(TestFileName, 0L, 10),
+                    new SendPacketsElement(stream, 0L, 0),
+                    new SendPacketsElement(TestFileName, 0L, 10),
+                };
+                var expected = GetExpectedContent(elements);
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.netcoreapp.cs]: Nit: var => ?

File: src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs (hunk 12 above)
Comment (kurtschelfthout) [reply src/System.Net.Sockets/tests/FunctionalTests/SendPacketsAsync.cs]: It's absolutely arbitrary. Just trying to limit the number of receives and test duration to a reasonable time before declaring failure. In practice the tests don't exceed this limit, and I reckon it's pretty unlikely to given they all execute locally.

All that said, happy to remove the `MaxTries` altogether if you prefer.

**Comment by kurtschelfthout** [comment]
test OSX x64 Debug Build

Catastrophic failure in a few unrelated test runs.

**Comment by kurtschelfthout** [comment]
Done with the re-re-review 😃

**Pull request merged**